docs/source/opener_parameters_api.md

```diff
@@ -0,0 +1,220 @@
+`min_datetime` and `max_datetime`. These properties must also conform to the
+`date` or `date-time` format.
+
+## Generating a UI from a schema
+
+With the addition of the time limits extension described above, the JSON
+Schema returned by `get_open_data_params_schema` is expected to be extensive
+and detailed enough to fully describe a UI for cube generation.
+
+### Order of properties in a schema
+
+Sub-elements of a `JsonObjectSchema` are passed to the constructor using the
+`properties` parameter with type signature `Mapping[str, JsonSchema]`. Openers
+SHOULD provide an ordered mapping as the value of `properties`, with the
+elements placed in an order suitable for presentation in a UI, and UI
+generators SHOULD lay out the UI in the provided order, with the exception of
+the common parameters discussed below. Note that the CPython `dict` object
+preserves the insertion order of its elements as of Python 3.6, and that this
+behaviour is officially guaranteed as of Python 3.7, so additional classes
+like `OrderedDict` are no longer necessary to fulfil this requirement.
+
+### Special handling of common parameters
+
+Any of the common parameters listed above SHOULD, if present, be recognized
+and handled specially. They SHOULD be presented in a consistent position
+(e.g. at the top of the page for a web GUI), in a consistent order, and with
+user-friendly labels and tooltips even if the `title` and `description`
+annotations (see below) are absent. The UI generator MAY provide special
+representations for these parameters, for instance an interactive map for the
+`bbox` parameter.
+
+An opener MAY provide `title`, `description`, and/or `examples` annotations
+for any of the common parameters, and a UI generator MAY choose to use any of
+these to supplement or modify its standard presentation of the common
+parameters.
+
+### Schema annotations (title, description, examples, and default)
+
+For JSON Schemas describing parameters other than the common parameters, an
+opener SHOULD provide the `title` and `description` annotations. A UI
+generator SHOULD make use of these annotations, for example by taking the
+label for a UI control from `title` and the tooltip from `description`. The
+opener and UI generator MAY additionally make use of the `examples` annotation
+to record and display example values for a parameter. If a sensible default
+value can be envisaged, the opener SHOULD record this default as the value of
+the `default` annotation and the UI generator SHOULD set the default value in
+the UI accordingly. If the `title` annotation is absent, the UI generator
+SHOULD use the key corresponding to the parameter's schema in the parent
+schema as a fallback.
+
+### Generalized conversion of parameter schemas
+
+For parameters other than the common parameters, the UI can be generated
+automatically from the schema structure. In the case of a GUI, a one-to-one
+conversion of instances of `JsonSchema` subclasses into GUI elements will
```

Comment:
I'd prefer a translation of _JSON schema_ to GUI elements here, rather than a translation of `JsonSchema` subclasses to GUI elements. In the GUI, we will receive plain JSON and we should make no assumption of how this JSON was generated.

Reply:
Agreed, and updated accordingly.